homeassistant/components/stream/worker.py

```diff
@@ -119,33 +144,71 @@ def mux_packet(self, packet):
             packet.stream = self._output_audio_stream
             self._av_output.mux(packet)
 
-    def flush(self, duration):
+    def check_flush_part(self, packet: av.Packet) -> None:
+        """Check for and mark a part segment boundary and record its duration."""
+        byte_position = self._memory_file.tell()
+        if self._segment.last_write_pos == byte_position:
+            return
+        if self._segment.last_write_pos == 0:  # The beginning of the first moof
+            self._segment.last_write_pos = byte_position
+            self._segment.init = self._memory_file.getbuffer()[:byte_position].tobytes()
+            # Fetch the latest StreamOutputs, which may have changed since the
+            # worker started.
+            self._outputs = self._outputs_callback().values()
+            for stream_output in self._outputs:
+                stream_output.put(self._segment)
+        else:  # These are the ends of the part segments
+            self._segment.parts.append(
+                Part(
+                    # The end of a part segment will actually have occurred before the last video packet
+                    duration=float(
+                        (self._last_packet_dts - self._part_start_dts)
```

Comment:
As mentioned above, each fragment should span several packets, so we are actually several packets in when this condition is called. The duration should not be zero.
The reason why we need `last_packet_dts` as opposed to `packet.pts` is actually a nuance which I tried to capture in the comment above the line. We've detected the file pointer moving, but that means that ffmpeg actually already received the first packet of the next fragment (that's how ffmpeg knows that the fragment is long enough), which means we muxed it already. Thus we should calculate the fragment duration off of that muxed packet, which is actually the last dts and not the current one.
If we assume all the `pyav` video `packet.duration` fields are filled in properly and that they are the same across time (constant framerate), we could just calculate the last packet's dts by subtracting the current packet.duration from the current packet.dts, so we can avoid keeping and updating the `last_packet_dts`. Is this what you meant?

Reply:
I'm still trying to wrap my head around the rest of this comment. Thanks for writing such a great detailed response.. I think it makes sense for me to spend the time to understand this rather than "rubber stamp" so hopefully you don't mind while i spend a little more time to catch up here.

Reply:
> Naively it seems like any time `flush` runs, it will have an empty `data` because `check_flush_part` already wrote it all. What stops that from happening?

So say the current packet is the `i`th packet. We base our decision of whether or not to close the segment on packet `i` (checking the segment duration and whether or not it is a keyframe).
However, closing happens before we mux this packet, so pyav is still on packet `i-1`.
`check_flush_part` checks whether a part has been produced by pyav by looking at the file pointer. The part boundaries from pyav are also determined by checking the packet duration, so if the part has been completed and written, it only contains packets up to packet `i-2`. Packet `i-1` will go in the next part and is still held internally by pyav. So even in the case where we have just done a part flush (flushing up to packet `i-2`) and we determine that packet `i` belongs in the next segment, we still have packet `i-1` left in pyav which needs to be written to a new fragment in the current segment. This "stub" fragment is not written until we close the `OutputContainer` which happens in `flush`.

Reply:
Is the `_memory_file` populated when the `self._av_output.mux` call happens or somewhere else by pyav before `mux_packet` is called? I was thinking if its when the output is mixed, then maybe the construction of the `Part`s should happen at the end of `mux_packet` instead rather than at the start.  I'm playing around structuring the code this way and it seems like less state to track, but I didn't try it in the code with actual tests running.

Reply:
If it results in a write I think it should happen right after mux_packet is called. I can try moving the Part construction down there.

Reply:
OK thank you for the detailed explanation. I think I may follow up to send a PR with a comment explaining what I think may be happening to test my understanding.